notifications/v1/client.go

```diff
@@ -6,90 +6,153 @@ import (
 	if err != nil {
 		return nil, errors.Wrap(err, "create GCP Pub/Sub client")
 	}
-	return &Client{
-		logger:       logger.Scoped("notification.ClientV1"),
-		subscription: client.Subscription(subscriptionID),
+	subscription := client.Subscription(opts.SubscriptionID)
+	subscription.ReceiveSettings = opts.ReceiveSettings
+	return &subscriber{
+		logger:       logger.Scoped("notification.subscriber"),
+		handlers:     opts.Handlers,
+		subscription: subscription,
+		state:        atomic.NewString(stateIdle),
 	}, nil
 }
 
-// ReceiveHandlers is a collection of receive handlers for each type of SAMS
-// notifications. If the handler of a notification is nil, the notification will
-// be acknowledged automatically without any processing.
-//
-// If a handler returns an error, the notification will be unacknowledged and
-// retried later.
-type ReceiveHandlers struct {
-	// OnUserDeleted is called when a "UserDeleted" notification is received.
-	OnUserDeleted func(data *UserDeletedData) error
-}
+func (s *subscriber) Start() {
+	if !s.state.CompareAndSwap(stateIdle, stateStarted) {
+		s.logger.Error("failed to start subscriber", log.Error(errors.New(`not in "idle" state`)))
+		return
+	}
 
-// Receive starts receiving SAMS notifications and calls the corresponding
-// handler for each notification. It blocks until the context is done (e.g.
-// deadline exceed or canceled).
-//
-// ⚠️ WARNING: Each subscription can only have one active receiver at a time,
-// i.e. there should only be one call to Receive for a given subscription.
-func (c *Client) Receive(handler *ReceiveHandlers) error {
 	ctx, cancel := context.WithCancel(context.Background())
-	c.cancelContext = cancel
-	return c.subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
+	s.cancelContext = cancel
+	err := s.subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		var msgData struct {
 			Name     string          `json:"name"`
 			Metadata json.RawMessage `json:"metadata"`
 		}
 		err := json.Unmarshal(msg.Data, &msgData)
 		if err != nil {
-			c.logger.Error("failed to unmarshal notification message", log.Error(err))
+			s.logger.Error("failed to unmarshal notification message", log.Error(err))
 			msg.Nack()
 			return
 		}
 
-		err = c.handleReceive(handler, msgData.Name, msgData.Metadata)
-		if err != nil {
-			c.logger.Error("failed to process notification message", log.Error(err))
+		err = s.handleReceive(msgData.Name, msgData.Metadata)
+		if err == nil {
+			msg.Ack()
+		} else {
+			s.logger.Error("failed to process notification message", log.Error(err))
 			msg.Nack()
-			return
 		}
-		msg.Ack()
 	})
+	if err != nil {
+		s.logger.Error("failed to receive notifications", log.Error(err))
+		return
+	}
+}
+
+func (r *subscriber) Stop() {
+	if !r.state.CompareAndSwap(stateStarted, stateStopped) {
+		r.logger.Error("failed to stop subscriber", log.Error(errors.New(`not in "started" state`)))
+		return
+	}
+
+	r.cancelContext()
```

Comment:
Should we have a `nil` check here in case `cancelContext` wasn't set? For example, on line 69 a `subscriber` is created _without_ specifying the cancel context. So if we are expecting this value to "come later", then just to err on the side of caution, then perhaps we should anticipate it not being set at all?

Though, thinking about it, I guess asserting that we are in the "stopped" state may be sufficient to ensure that `Start()` was called... and perhaps that is where `cancelContext` would be set?

Reply:
I think the state check is sufficient, though it may help to have a docstring here indicating as such

Reply:
Good catch, tho I'd generally say state check is enough here, but I pushed a commit https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/75628f4861c26807eae94b5f4e37347d6f9fc0c8 that always gives `cancelContext` a default complaining it is called without being set (with a useful value) as the simple safeguard.